File: java-app/src/main/java/com/showcase/java/slice/nullsafety/NullSafeWorkflow.java
package com.showcase.java.slice.nullsafety;

import com.showcase.java.domain.event.DomainEvent;
import com.showcase.java.domain.model.*;

import java.time.LocalDate;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Demonstrates Java 25's null handling patterns in an event-sourced task workflow.
 *
 * <p>Key Java 25 features showcased:
 * <ul>
 *   <li>{@link Optional} for explicit nullable return types</li>
 *   <li>{@code @Nullable} annotations (documentation-level, not compiler-enforced)</li>
 *   <li>Null-pattern in switch expressions ({@code case null ->})</li>
 *   <li>Pattern matching for null checks</li>
 * </ul>
 *
 * <p>Comparison with Kotlin: Java 25 relies on runtime checks and conventions.
 * {@code @Nullable} annotations are advisory — the compiler does not reject null
 * assignments to non-annotated parameters. {@link Optional} adds ceremony but
 * provides explicit intent. The null-pattern in switch is a nice ergonomic
 * improvement but does not provide compile-time guarantees.
 */
public class NullSafeWorkflow {

    private final Map<String, TaskSummary> tasks = new ConcurrentHashMap<>();

    // -----------------------------------------------------------------------
    // OPTIONAL FOR NULLABLE RETURNS
    // -----------------------------------------------------------------------

    /**
     * Looks up a task and returns its assignee name using Optional chains.
     *
     * <p>Java uses {@link Optional} to signal that a value may be absent.
     * The chain {@code .map().map()} mirrors Kotlin's safe call chain {@code ?.}.
     */
    public Optional<String> findAssigneeName(String taskId) {
        return Optional.ofNullable(tasks.get(taskId))
                .map(TaskSummary::assignee)
                .map(UserId::value);
    }

    /**
     * Returns the due date of a task, or empty if the task doesn't exist or has no due date.
     */
    public Optional<LocalDate> findDueDate(String taskId) {
        return Optional.ofNullable(tasks.get(taskId))
                .map(TaskSummary::dueDate);
    }

    // -----------------------------------------------------------------------
    // DEFAULT VALUES (orElse — equivalent to Kotlin's elvis ?:)
    // -----------------------------------------------------------------------

    /**
     * Returns the assignee name or a default placeholder.
     *
     * <p>{@code orElse} is Java's equivalent of Kotlin's elvis operator {@code ?:}.
     */
    public String getAssigneeOrDefault(String taskId) {
        return Optional.ofNullable(tasks.get(taskId))
                .map(TaskSummary::assignee)
                .map(UserId::value)
                .orElse("unassigned");
    }

    /**
     * Returns the task title or throws if the task doesn't exist.
     *
     * <p>{@code orElseThrow} is Java's equivalent of Kotlin's {@code ?: throw}.
     */
    public String getTaskTitleOrThrow(String taskId) {
        return Optional.ofNullable(tasks.get(taskId))
                .map(TaskSummary::title)
                .orElseThrow(() -> new NoSuchElementException("Task " + taskId + " not found"));
    }

    // -----------------------------------------------------------------------
    // NULL-PATTERN SWITCH (Java 25)
    // -----------------------------------------------------------------------

    /**
     * Demonstrates Java 25's null-pattern in switch for handling nullable task lookups.
     *
     * <p>The {@code case null} arm handles the null case explicitly in the switch,
     * avoiding a separate null check before the switch.
     */
    public String describeTask(String taskId) {
        var task = tasks.get(taskId);
        return switch (task) {
            case null -> "Task " + taskId + " not found";
            case TaskSummary t -> "Task: " + t.title() + " [" + describeStatus(t.status()) + "]";
        };
    }

    /**
     * Describes a task status using exhaustive switch with null-pattern support.
     */
    public String describeStatus(TaskStatus status) {
        return switch (status) {
            case null -> "unknown";
            case TaskStatus.Open _ -> "open";
            case TaskStatus.InProgress s -> "in progress, assigned to " + s.assignee().value();
            case TaskStatus.Blocked s -> "blocked: " + s.reason();
            case TaskStatus.Completed s -> "completed at " + s.completedAt();
        };
    }

    // -----------------------------------------------------------------------
    // ifPresent — equivalent to Kotlin's ?.let { }
    // -----------------------------------------------------------------------

    /**
     * Processes a task assignment using Optional.ifPresent for conditional execution.
     *
     * <p>This mirrors Kotlin's {@code ?.let { ... }} pattern.
     */
    public Optional<TaskSummary> processAssignment(String taskId, UserId assigneeId) {
        var existing = tasks.get(taskId);
        if (existing == null) {
            return Optional.empty();
        }
        var updated = new TaskSummary(
                existing.id(),
                existing.title(),
                new TaskStatus.InProgress(assigneeId),
                existing.priority(),
                assigneeId,
                existing.dueDate(),
                existing.tags()
        );
        tasks.put(taskId, updated);
        return Optional.of(updated);
    }

    // -----------------------------------------------------------------------
    // EVENT REPLAY WITH NULL HANDLING
    // -----------------------------------------------------------------------

    /**
     * Replays events to build task state, demonstrating Java null handling patterns.
     *
     * <p>Java has no compile-time null safety — nullable fields like {@code description}
     * and {@code dueDate} in records are just regular references that may be null.
     * The programmer must remember to check; the compiler won't help.
     */
    public void replayEvents(List<DomainEvent> events) {
        for (var event : events) {
            switch (event) {
                case DomainEvent.TaskCreated e -> tasks.put(e.aggregateId(), new TaskSummary(
                        e.aggregateId(),
                        e.title(),
                        new TaskStatus.Open(),
                        e.priority(),
                        null,        // no assignee yet — null, not Optional
                        e.dueDate(), // may be null — no compile-time warning
                        e.tags()
                ));

                case DomainEvent.TaskAssigned e -> {
                    var existing = tasks.get(e.aggregateId());
                    if (existing != null) {
                        tasks.put(e.aggregateId(), new TaskSummary(
                                existing.id(),
                                existing.title(),
                                new TaskStatus.InProgress(e.assigneeId()),
                                existing.priority(),
                                e.assigneeId(),
                                existing.dueDate(),
                                existing.tags()
                        ));
                    }
                }

                case DomainEvent.TaskCompleted e -> {
                    var existing = tasks.get(e.aggregateId());
                    if (existing != null) {
                        tasks.put(e.aggregateId(), existing.withStatus(
                                new TaskStatus.Completed(e.completedAt())
                        ));
                    }
                }

                case DomainEvent.TaskPriorityChanged e -> {
                    var existing = tasks.get(e.aggregateId());
                    if (existing != null) {
                        tasks.put(e.aggregateId(), existing.withPriority(e.newPriority()));
                    }
                }

                case DomainEvent.TagAddedToTask e -> {
                    var existing = tasks.get(e.aggregateId());
                    if (existing != null) {
                        var newTags = new HashSet<>(existing.tags());
                        newTags.add(e.tag());
                        tasks.put(e.aggregateId(), existing.withTags(Set.copyOf(newTags)));
                    }
                }

                case DomainEvent.TaskMovedToProject _ -> {
                    // No status change
                }
            }
        }
    }

    /**
     * Returns all tasks currently tracked.
     */
    public List<TaskSummary> getAllTasks() {
        return List.copyOf(tasks.values());
    }
}

File: kotlin-app/src/main/kotlin/com/showcase/kotlin/slice/nullsafety/NullSafeWorkflow.kt
package com.showcase.kotlin.slice.nullsafety

import com.showcase.kotlin.domain.event.DomainEvent
import com.showcase.kotlin.domain.model.*
import java.time.Instant
import java.time.LocalDate

/**
 * Demonstrates Kotlin's compile-time null safety in an event-sourced task workflow.
 *
 * Key Kotlin features showcased:
 * - Nullable vs non-nullable types enforced at compile time
 * - Safe call operator `?.`
 * - Elvis operator `?:`
 * - Scope functions `let` and `also` for null handling
 * - Compile-time rejection of unsafe null operations (shown in comments)
 *
 * Parity Level: "Kotlin still leads" — Kotlin enforces null safety at compile time,
 * while Java 25 relies on runtime checks, Optional, and annotations.
 */
class NullSafeWorkflow(
    private val tasks: MutableMap<String, TaskSummary> = mutableMapOf()
) {

    // -----------------------------------------------------------------------
    // COMPILE-TIME REJECTION EXAMPLES
    // The following lines would NOT compile in Kotlin. They are commented out
    // to show what the compiler rejects.
    // -----------------------------------------------------------------------

    // ERROR: Null can not be a value of a non-null type String
    // val title: String = null

    // ERROR: Only safe (?.) or non-null asserted (!!) calls are allowed
    //        on a nullable receiver of type String?
    // val len: Int = nullableString.length

    // ERROR: Type mismatch — required: UserId, found: UserId?
    // fun assignTask(assignee: UserId?) { val id: UserId = assignee }

    // -----------------------------------------------------------------------
    // SAFE CALL CHAINS (?.)
    // -----------------------------------------------------------------------

    /**
     * Looks up a task and returns its assignee name using safe call chains.
     *
     * `tasks[taskId]` returns `TaskSummary?` (nullable).
     * `?.assignee` safely accesses the nullable assignee field.
     * `?.value` safely accesses the UserId value.
     * If any link in the chain is null, the entire expression evaluates to null.
     */
    fun findAssigneeName(taskId: String): String? =
        tasks[taskId]?.assignee?.value

    /**
     * Returns the due date of a task, or null if the task doesn't exist or has no due date.
     * Demonstrates chained safe calls on nullable properties.
     */
    fun findDueDate(taskId: String): LocalDate? =
        tasks[taskId]?.dueDate

    // -----------------------------------------------------------------------
    // ELVIS OPERATOR (?:)
    // -----------------------------------------------------------------------

    /**
     * Returns the assignee name or a default placeholder.
     *
     * The elvis operator `?:` provides a fallback when the left side is null.
     * This replaces verbose if-null-then-else patterns.
     */
    fun getAssigneeOrDefault(taskId: String): String =
        tasks[taskId]?.assignee?.value ?: "unassigned"

    /**
     * Returns the task title or throws if the task doesn't exist.
     *
     * Elvis can also be used with `throw` for fail-fast behavior.
     */
    fun getTaskTitleOrThrow(taskId: String): String =
        tasks[taskId]?.title ?: throw NoSuchElementException("Task $taskId not found")

    // -----------------------------------------------------------------------
    // SCOPE FUNCTIONS FOR NULL HANDLING (let, also)
    // -----------------------------------------------------------------------

    /**
     * Processes a task assignment using `let` for null-safe transformation.
     *
     * `?.let { ... }` executes the block only if the receiver is non-null.
     * Inside the lambda, `it` is guaranteed non-null — no further null checks needed.
     */
    fun processAssignment(taskId: String, assigneeId: UserId): TaskSummary? =
        tasks[taskId]?.let { task ->
            // `task` is guaranteed non-null here — smart cast by `let`
            val updated = task.copy(
                assignee = assigneeId,
                status = TaskStatus.InProgress(assignee = assigneeId)
            )
            tasks[taskId] = updated
            updated
        }

    /**
     * Looks up a task and logs it using `also` for side effects.
     *
     * `?.also { ... }` executes the side-effect block only if non-null,
     * then returns the original value unchanged.
     */
    fun findTaskWithLogging(taskId: String, logger: (String) -> Unit): TaskSummary? =
        tasks[taskId]?.also { task ->
            logger("Found task: ${task.title} [${task.status}]")
        }

    // -----------------------------------------------------------------------
    // COMBINED NULL SAFETY IN EVENT REPLAY
    // -----------------------------------------------------------------------

    /**
     * Replays events to build task state, demonstrating combined null safety patterns.
     *
     * - Non-nullable parameters: `events` list and each `DomainEvent` are non-null by type
     * - Nullable fields: `description`, `dueDate` are nullable in TaskCreated
     * - Safe calls + elvis for lookups during replay
     * - `let` for conditional updates on nullable task lookups
     */
    fun replayEvents(events: List<DomainEvent>) {
        for (event in events) {
            when (event) {
                is DomainEvent.TaskCreated -> {
                    tasks[event.aggregateId] = TaskSummary(
                        id = event.aggregateId,
                        title = event.title,
                        status = TaskStatus.Open,
                        priority = event.priority,
                        assignee = null,       // nullable — no assignee yet
                        dueDate = event.dueDate, // nullable — may or may not have a due date
                        tags = event.tags
                    )
                }

                is DomainEvent.TaskAssigned -> {
                    // Safe call + let: only update if the task exists
                    tasks[event.aggregateId]?.let { existing ->
                        tasks[event.aggregateId] = existing.copy(
                            assignee = event.assigneeId,
                            status = TaskStatus.InProgress(assignee = event.assigneeId)
                        )
                    }
                }

                is DomainEvent.TaskCompleted -> {
                    tasks[event.aggregateId]?.let { existing ->
                        tasks[event.aggregateId] = existing.copy(
                            status = TaskStatus.Completed(completedAt = event.completedAt)
                        )
                    }
                }

                is DomainEvent.TaskPriorityChanged -> {
                    tasks[event.aggregateId]?.let { existing ->
                        tasks[event.aggregateId] = existing.copy(priority = event.newPriority)
                    }
                }

                is DomainEvent.TagAddedToTask -> {
                    tasks[event.aggregateId]?.let { existing ->
                        tasks[event.aggregateId] = existing.copy(tags = existing.tags + event.tag)
                    }
                }

                is DomainEvent.TaskMovedToProject -> {
                    // No status change — task stays in current state
                }
            }
        }
    }

    /**
     * Returns all tasks currently tracked.
     */
    fun getAllTasks(): List<TaskSummary> = tasks.values.toList()
}
